File: zad2_3/app/src/main/java/com/pam/bmi/ui/quiz/QuizFragment.java
package com.pam.bmi.ui.quiz;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.pam.bmi.R;

import androidx.fragment.app.Fragment;

import com.pam.bmi.ui.models.Question;

import java.util.ArrayList;

public class QuizFragment extends Fragment {
    private TextView mQuestionTextView;
    private RadioGroup mRadioGroup;
    private RadioButton mRadioButton1;
    private RadioButton mRadioButton2;
    private RadioButton mRadioButton3;
    private Button mNextButton;

    private ArrayList<Question> questionList;
    private int mCurrentIndex;
    private int mScore;

    private View view;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_quiz, container, false);

        initializeViews();
        setOnClickListeners();

        initiateQuestions();
        mCurrentIndex = 0;
        mScore = 0;
        setQuestion();

        return view;
    }

    private void initializeViews() {
        mQuestionTextView = view.findViewById(R.id.question_text_view);
        mRadioGroup = view.findViewById(R.id.answer_radio_group);
        mRadioButton1 = view.findViewById(R.id.choice_a_button);
        mRadioButton2 = view.findViewById(R.id.choice_b_button);
        mRadioButton3 = view.findViewById(R.id.choice_c_button);
        mNextButton = view.findViewById(R.id.submit_button);
    }

    private void setOnClickListeners() {
        mNextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                handleNextButtonClick();
            }
        });
    }

    private void handleNextButtonClick() {
        checkAnswer();
        mCurrentIndex++;
        if (mCurrentIndex < questionList.size()) {
            setQuestion();
        } else {
            endQuiz();
        }
    }

    private void setQuestion() {
        Question currentQuestion = questionList.get(mCurrentIndex);

        mQuestionTextView.setText(currentQuestion.getQuestion());
        mRadioButton1.setText(currentQuestion.getAnswers()[0]);
        mRadioButton2.setText(currentQuestion.getAnswers()[1]);
        mRadioButton3.setText(currentQuestion.getAnswers()[2]);
        mRadioGroup.clearCheck();
    }

    private void checkAnswer() {
        int correctAnswerIndex = questionList.get(mCurrentIndex).getCorrectAnswerIndex();
        int selectedAnswerIndex = mRadioGroup.getCheckedRadioButtonId();

        if (selectedAnswerIndex != -1) {
            RadioButton selectedAnswer = view.findViewById(selectedAnswerIndex);
            int selectedAnswerValue = mRadioGroup.indexOfChild(selectedAnswer);
            if (selectedAnswerValue == correctAnswerIndex) {
                mScore++;
            }
        }
    }

    private void endQuiz() {
        String text = "Your score is " + mScore + " out of " + questionList.size();
        showEndQuizDialog(text);
    }

    private void showEndQuizDialog(String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setMessage(message)
                .setCancelable(false)
                .setPositiveButton("Play Again", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        mCurrentIndex = 0;
                        mScore = 0;
                        setQuestion();
                    }
                });

        AlertDialog alert = builder.create();
        alert.show();
    }

    private void initiateQuestions() {
        questionList = new ArrayList<>();
        for (int i = 1; i <= 5; i++) {
            Question question = new Question(
                    getResources().getString(getResources().getIdentifier("question_" + i, "string", getActivity().getPackageName())),
                    new String[]{
                            getResources().getString(getResources().getIdentifier("q" + i + "_answer_a", "string", getActivity().getPackageName())),
                            getResources().getString(getResources().getIdentifier("q" + i + "_answer_b", "string", getActivity().getPackageName())),
                            getResources().getString(getResources().getIdentifier("q" + i + "_answer_c", "string", getActivity().getPackageName()))
                    },
                    Integer.parseInt(getResources().getString(getResources().getIdentifier("q" + i + "_answer_correct_index", "string", getActivity().getPackageName())))
            );
            questionList.add(question);
        }
    }
}

File: zad_4_kotlin/app/src/main/java/com/pam/bmi/ui/quiz/QuizFragment.kt
package com.pam.bmi.ui.quiz

import android.app.AlertDialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.pam.bmi.R
import com.pam.bmi.ui.models.Question

class QuizFragment : Fragment() {
    private lateinit var mQuestionTextView: TextView
    private lateinit var mRadioGroup: RadioGroup
    private lateinit var mRadioButton1: RadioButton
    private lateinit var mRadioButton2: RadioButton
    private lateinit var mRadioButton3: RadioButton
    private lateinit var mNextButton: Button
    private lateinit var questionList: ArrayList<Question>
    private var mCurrentIndex = 0
    private var mScore = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_quiz, container, false)
        initializeViews(view)
        setOnClickListeners()
        initiateQuestions()
        mCurrentIndex = 0
        mScore = 0
        setQuestion()
        return view
    }

    private fun initializeViews(view: View) {
        mQuestionTextView = view.findViewById(R.id.question_text_view)
        mRadioGroup = view.findViewById(R.id.answer_radio_group)
        mRadioButton1 = view.findViewById(R.id.choice_a_button)
        mRadioButton2 = view.findViewById(R.id.choice_b_button)
        mRadioButton3 = view.findViewById(R.id.choice_c_button)
        mNextButton = view.findViewById(R.id.submit_button)
    }

    private fun setOnClickListeners() {
        mNextButton.setOnClickListener { handleNextButtonClick() }
    }

    private fun handleNextButtonClick() {
        checkAnswer()
        mCurrentIndex++
        if (mCurrentIndex < questionList.size) {
            setQuestion()
        } else {
            endQuiz()
        }
    }

    private fun setQuestion() {
        val currentQuestion = questionList[mCurrentIndex]
        mQuestionTextView.text = currentQuestion.question
        mRadioButton1.text = currentQuestion.answers[0]
        mRadioButton2.text = currentQuestion.answers[1]
        mRadioButton3.text = currentQuestion.answers[2]
        mRadioGroup.clearCheck()
    }

    private fun checkAnswer() {
        val correctAnswerIndex = questionList[mCurrentIndex].correctAnswerIndex
        val selectedAnswerIndex = mRadioGroup.checkedRadioButtonId
        if (selectedAnswerIndex != -1) {
            val selectedAnswer = view?.findViewById<RadioButton>(selectedAnswerIndex)
            val selectedAnswerValue = mRadioGroup.indexOfChild(selectedAnswer)
            if (selectedAnswerValue == correctAnswerIndex) {
                mScore++
            }
        }
    }

    private fun endQuiz() {
        val text = "Your score is $mScore out of ${questionList.size}"
        showEndQuizDialog(text)
    }

    private fun showEndQuizDialog(message: String) {
        val builder = AlertDialog.Builder(context)
        builder.setMessage(message)
                .setCancelable(false)
                .setPositiveButton("Play Again") { _, _ ->
                    mCurrentIndex = 0
                    mScore = 0
                    setQuestion()
                }
        val alert = builder.create()
        alert.show()
    }

    private fun initiateQuestions() {
        questionList = ArrayList()
        for (i in 1..6) {
            val question = Question(
                    resources.getString(resources.getIdentifier("question_$i", "string", requireActivity().packageName)), arrayOf(
                    resources.getString(resources.getIdentifier("q" + i + "_answer_a", "string", requireActivity().packageName)),
                    resources.getString(resources.getIdentifier("q" + i + "_answer_b", "string", requireActivity().packageName)),
                    resources.getString(resources.getIdentifier("q" + i + "_answer_c", "string", requireActivity().packageName))
            ), resources.getInteger(resources.getIdentifier("q" + i + "_answer_correct_index", "integer", requireActivity().packageName))
            )
            questionList.add(question)
        }
    }
}
